### ISSUE TITLE
Updating chart stops following last value

### ISSUE THREAD

**Issue by iNikNik** [rrag/react-stockcharts#70]
Hello! There is a little bug - updating chart stops following last value. It happens randomly. If you run ["updating data" example](http://rrag.github.io/react-stockcharts/documentation.html#/updating_data) on max speed (update speed is not the reason) and wait - you'll see it.

I've spent some time debugging this and I've figured out the issue: [getDataBetween](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L18) calculates wrong [newRightIndex](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L22). Maybe [getClosestItemIndexes ](https://github.com/rrag/react-stockcharts/blob/master/src/lib/utils/index.js#L37) is the reason. I don't really understand why we need this calculations, but I've found fix: change it
```javascript
var newLeftIndex = getClosestItemIndexes(dataForInterval, left, xAccessor).right;
var newRightIndex = getClosestItemIndexes(dataForInterval, right, xAccessor).left;
```
to 
```javascript
var newLeftIndex = getClosestItemIndexes(dataForInterval, left, xAccessor).left;
var newRightIndex = getClosestItemIndexes(dataForInterval, right, xAccessor).right;
```

Also, I've created [this fiddle](http://jsfiddle.net/f8fg8qa6/1/) with my data. So, you can reproduce the issue. getDataBetween've called form https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L160 . There is array[364] and last index should be 363 but left == 362 and at the next update last item will become invisible.

Should I create PR for this?

**Comment by rrag** [comment]
if 
```
right = 362.99999999999994
```
then the chart should show up to index `362` since `363 > 362.99999999999994` it makes sense that it does not show `363`

in the example when the updating stops, the last candle is `2015-06-08` which is the last value in the `MSFT_full.tsv` file.

In your case somehow `right` became `362.99999999999994` instead of `363`, that is the real problem. I guess it is because of something like `.1+.2 = 0.30000000000000004` in javascript.

May be we need to round the index to nearest 2 decimal places, which could solve this. Thoughts?

**Comment by iNikNik** [comment]
> in the example when the updating stops, the last candle is 2015-06-08 which is the last value in the MSFT_full.tsv file.

It's hard to reproduce this issue. Today i've tried couple times with no success (on `updating data` example). But the goal is - it stops following last value **before** `2015-06-08`.

I've recorded gif to show you what is happening:
![chart_issue](https://cloud.githubusercontent.com/assets/7679749/14707966/51e3ee58-07d0-11e6-8c79-18769a8d11be.gif)

In this case - it randomly stops following last value. But if I pan right - new values are here. Here is another gif with log:
![chart_issue2](https://cloud.githubusercontent.com/assets/7679749/14708063/098ab906-07d1-11e6-9aed-71df4037800c.gif)

You can see that before it says `DATA CHANGED AND LAST ITEM VISIBLE` but after `TRIVIAL CHANGE`.

So, like i said before there is some issue :( 

> In your case somehow right became 362.99999999999994 instead of 363, that is the real problem. I guess it is because of something like .1+.2 = 0.30000000000000004 in javascript.

Yes, maybe here https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L158 we're getting bad values, but I don't see how we can fix it here. 
Maybe we should round [value here](.toFixed(4)) like this:
```javascript
var [newStart, newEnd] = updatedXScale
  .range()
  .map(x => x + dx)
  .map(updatedXScale.invert)
  .map(x => Math.round(x* 10000) / 10000)
```
That will round 362.99999999999994 to 363.0000

**Comment by iNikNik** [comment]
Yep, it fixed the issue.

**Comment by rrag** [comment]
Exactly what I was thinking.

But the rounding has to be applied conditionally only if a `discontinuous` scale is used. The library supports any d3 scale. so if time scale is used that rounding will not work.

It might be ok for your fork if you do not use any other scale.

**Comment by iNikNik** [comment]
@rrag can't understand why it'll not work with time scale. `newStart, newEnd` are indexes, right? I mean they are the numbers and when we're adding dx - `x => x + dx` we can do some rounding. How it can break scale? But if it does - we need to find other solution.

**Comment by rrag** [comment]
```js
var [newStart, newEnd] = updatedXScale
  .range()
  .map(x => x + dx) // here we are operating on range (usually 0 - chartwidth) which will always be a number
  .map(updatedXScale.invert) // range is inverted here, it becomes domain now. so depending on the scale, say d3.time.scale() the domain will be Date, if the scale is a linearscale like the financeEODScale, then the domain will be number
  .map(x => Math.round(x* 10000) / 10000) // if d3.time.scale then it would round a date value
```

**Comment by iNikNik** [comment]
@rrag ooh... it makes sense now. Such a dumb comment from me) I'm not familiar with d3 yet.

So, I believe we can safely put it inside [invert  method](https://github.com/rrag/react-stockcharts/blob/master/src/lib/scale/financeEODDiscontiniousScale.js#L55-L57)
```javascript
scale.invert = function(x) {
    return Math.round(backingLinearScale.invert(x) * 10000) / 10000;
};
```
It'll fix the issue

**Comment by rrag** [comment]
Added this fix in the [new scale](https://github.com/rrag/react-stockcharts/blob/dcfc1b4426341cdcb478213a96b9cb14f2d1b96d/src/lib/scale/financeDiscontinuousScale.js#L40-L41) will be published as part of 0.5
